### ISSUE TITLE
allow_nil option for validate_uniqueness_of

### ISSUE THREAD

**Issue by JulianKniephoff** [thoughtbot/shoulda#129]
I added an option that checks whether the validation of uniqueness is skipped, when the attribute in question has a value of nil. So basically it tries to enforce something along the lines of

```
validates_uniqueness_of :attr, :allow_nil => true # (a)
```

The branch can be found here: http://github.com/JulianKniephoff/shoulda/tree/validate_uniqueness_of_allow_nil

validate_uniqueness_of with allow_nil option given now matches models having such a clause in them. It does not match a model with just

```
validates_uniqueness_of :attr # (b)
```

as it does not allow nil duplicates. A class with statement (a) on the other hand is no longer matched by just validate_uniqueness_of (without allow_nil option) as it allows a duplicate: nil. Accordingly, validates_uniqueness_of without allow_nil _does_ (still) match a model with (b) in it.

In summary:

<table>
<thead>
<tr>
<td><code>validates_uniqueness_of :attr, :allow_nil => true</code> (Model)</td>
<td><code>validate_uniqueness_of(:attr).allow_nil</code> (Test)</td>
<td>matches?</td>
</tr>
</thead>
<tbody>
<tr>
<td>no</td>
<td>no</td>
<td>yes</td>
</tr>
<tr>
<td>no</td>
<td>yes</td>
<td>no</td>
</tr>
<tr>
<td>yes</td>
<td>no</td>
<td>no</td>
</tr>
<tr>
<td>yes</td>
<td>yes</td>
<td>yes</td>
</tr>
</tbody>
</table>

I hope this behavior conforms with Shoulda.
